### PR TITLE
Implement compileHistories() for UpdatedType:: DYNAMI for Hdf5 Recorder

### LINKED ISSUE #658: Implement Hdf5Recorder::initDataset() and compileHistories() for UpdatedType:: DYNAMIC
This will allow us to create HDF5 files that hold data that is generated incrementally during simulation and saved after each simulation epoch.

### PR DESCRIPTION
**Pull request by Hsu0227**
Closes #658 

#### Description
Implement compileHistories() for UpdatedType:: DYNAMIC for Hdf5 Recorder, inside the compileHistories(), in order to allow us to create HDF5 files that hold data that is generated incrementally during the simulation and saved after each simulation epoch, we need to extend the size of the dataset, so I change a way to create a dataset by using chunk, and then using the hyperslab to the extended portion of the dataset, but don't forget to call startNewEpoch at the beginning of each simulation epoch to prepare for recording new events.

### CODE REVIEW

File: Simulator/Recorders/Hdf5Recorder.cpp [hunk 1]
```diff
@@ -122,6 +123,94 @@ void Hdf5Recorder::saveSimData(const AllVertices &neurons)
    }
 }
 
+// Processes and updates HDF5 datasets for variables marked as DYNAMIC
+void Hdf5Recorder::compileHistories(AllVertices &vertices)
```
Comment (stiber) [inline Simulator/Recorders/Hdf5Recorder.cpp:127]: Does this method still need this parameter?

File: Simulator/Recorders/Hdf5Recorder.cpp [hunk 2]
```diff
@@ -122,6 +123,94 @@ void Hdf5Recorder::saveSimData(const AllVertices &neurons)
...
+{
+   // Define the maximum chunk size for datasets to optimize storage and access
+   const hsize_t max_chunk_size = 1024;
```
Comment (stiber) [inline Simulator/Recorders/Hdf5Recorder.cpp:130]: What happens if the amount of data is bigger? Good to put a comment about that.

File: Simulator/Recorders/Hdf5Recorder.cpp [hunk 3]
```diff
@@ -122,6 +123,94 @@ void Hdf5Recorder::saveSimData(const AllVertices &neurons)
...
+               } else if (variableInfo.hdf5Datatype_ == PredType::NATIVE_UINT64) {
+                  std::vector<uint64_t> dataBuffer(variableInfo.variableLocation_.getNumElements());
+                  for (size_t i = 0; i < variableInfo.variableLocation_.getNumElements(); ++i) {
+                     dataBuffer[i] = get<uint64_t>(variableInfo.variableLocation_.getElement(i));
+                  }
+                  variableInfo.hdf5DataSet_.write(dataBuffer.data(), variableInfo.hdf5Datatype_,
+                                                  memSpace, fileSpace);
+               } else {
+                  // Throw an exception if the data type is unsupported
+                  throw std::runtime_error("Unsupported data type");
```
Comment (stiber) [inline Simulator/Recorders/Hdf5Recorder.cpp:203]: Can you put a bit more information in the message? At least include the variable name.

File: Simulator/Recorders/Hdf5Recorder.h [hunk 4]
```diff
@@ -60,9 +60,7 @@ class Hdf5Recorder : public Recorder {
 
    // TODO: No parameters needed (AllVertices &vertices)
    /// Compile/capture variable history information in every epoch
-   virtual void compileHistories(AllVertices &neurons) override
-   {
-   }
+   virtual void compileHistories(AllVertices &neurons) override;
```
Comment (stiber) [inline Simulator/Recorders/Hdf5Recorder.h:63]: Remember to eliminate unnecessary parameters.

**Review by stiber — CHANGES_REQUESTED** [review]

File: Simulator/Recorders/Hdf5Recorder.cpp (hunk 1 above)
Comment (Hsu0227) [reply Simulator/Recorders/Hdf5Recorder.cpp:127]: no, they don't need parameter any more

File: Simulator/Recorders/Hdf5Recorder.h (hunk 4 above)
Comment (Hsu0227) [reply Simulator/Recorders/Hdf5Recorder.h:63]: yeah, I did it

**Review by stiber — APPROVED** [review]
Good job!

**Pull request merged**